### ISSUE TITLE
How to save fp16 features by buffers.getHostBuffer?

### ISSUE THREAD

**Issue by Bonsen** [NVIDIA/TensorRT#1180]
If I want to save the int8 feature, i can use:
.....
config->setFlag(BuilderFlag::kINT8);
.....
const int8_t* output = static_cast<int8_t*>(buffers.getHostBuffer(_additional_output));
std::vector\<int> embedding;
embedding.insert(embedding.end(), output, output + output_size_tr);

If I want to save the float32 feature, i can use:
const float* output = static_cast<float*>(buffers.getHostBuffer(_additional_output));
std::vector\<float> embedding;
embedding.insert(embedding.end(), output, output + output_size_tr);

How to save fp16 feature and if my int8 saving embedding right?
.....
config->setFlag(BuilderFlag::kFP16);
.....

**Comment by Bonsen** [comment]
@ttyio Do u know it?

**Comment by ttyio** [comment]
@Bonsen ,
we can set the `kSTRICT_TYPES` in builderflag, and mark the output using `ITensor::setType(kHALF)`, you could refer sample for more detail
https://github.com/NVIDIA/TensorRT/tree/release/7.2/samples/opensource/sampleReformatFreeIO

**Comment by Bonsen** [comment]
@ttyio 
----
I can get output from:
"_additional_output" isthe name of  global-avg-pool's output.
const int8_t* output = static_cast<const int8_t*>(buffers.getHostBuffer(_additional_output));
ector<float>& embedding = results_.classification_embedding;
embedding.insert(embedding.end(), output, output + output_size_tr);
cout << embedding << endl;
-----

when i set:
config->setFlag(BuilderFlag::kINT8);
layer->getOutput(0)->setType(nvinfer1::DataType::kINT8);
auto output = layer->getOutput(0);
network->markOutput(*output);

the output is: [3, 6, 6, 11, 0, 3, 0, 13, 14, 0, 9, 0, 9, 0, 14, 0, 2, 61, 15, 1, 8, 1, 0, 2, 2, 11, 0, 1, 0, 12, 1, 0, 1, 13, 4, 5, 1, 1, 1, 5, 1, 8, 24, 8, 0, 8, 0, 91, 0, 0, 1, 7, 0, 18, 1, 1, 0, 3, 0, 0, 25, 0, 8, 35, 0, 0, 2, 4, 19, 2, 8, 0, 0, 1, 4, 15, 18, 1, 1, 0, 0, 5, 4, 10, 1, 0, 1, 13, 2, 8, 30, 0, 7, 1, 4, 0, 0, 6, 7, 7, 0, 1, 2, 2, 0, 5, 14, 3, 0, 6, 8, 9, 29, 1, 1, 8, 0, 9, 0, 2, 7, 1, 0, 0, 48, 0, 11, 0, 34, 1, 2, 10, 0, 10, 0, 1, 16, 1, 0, 13, 0, 5, 12, 2, 7, 0, 31, 6, 0, 46, 1, 1, 4, 3, 28, 3, 28, 12, 0, 7, 4, 0, 3, 0, 1, 0, 0, 0, 1, 2, 2, 45, 3, 17, 9, 26, 2, 2, 62, 4, 0, 3, 0, 0, 3, 0, 7, 1, 0, 0, 25, 55, 2, 21, 0, 0, 1, 1, 0, 9, 0, 0, 1, 0, 9, 4, 5, 0, 3, 10, 0, 0, 4, 0, 1, 8, 0, 0, 52, 2, 0, 1, 44, 0, 0, 1, 5, 5, 17, 16, 25, 5, 13, 0, 46, 22, 0, 2, 0, 30, 2, 0, 11, 13, 4, 24, 46, 31, 4, 2, 9, 32, 1, 0, 0, 5, 8, 0, 0, 0, 17, 4, 1, 1, 1, 0, 0, 4, 17, 0, 1, 1, 10, 1, 43, 7, 38, 0, 3, 2, 0, 0, 0, 1, 5, 1, 2, 1, 2, 1, 8, 75, 7, 0, 11, 1, 8, 0, 1, 4, 2, 8, 114, 5, 0, 9, 1, 2, 4, 0, 2, 11, 1, 0, 3, 4, 0, 25, 6, 8, 69, 3, 1, 18, 9, 0, 0, 2, 9, 0, 1, 8, 1, 7, 5, 2, 10, 22, 34, 11, 1, 3, 0, 22, 10, 1, 9, 2, 0, 7, 5, 0, 12, 1, 3, 7, 3, 3, 27, 18, 3, 1, 6, 13, 67, 3, 7, 6, 0, 6, 43, 0, 0, 84, 1, 1, 14, 17, 5, 0, 0, 53, 6, 39, 12, 11, 1, 0, 1, 0, 0, 0, 7, 31, 1, 7, 0, 2, 1, 0, 32, 3, 3, 0, 3, 4, 77, 28, 0, 14, 6, 0, 0, 1, 0, 0, 0, 11......]
all >= 0??????????????

when i set:
config->setFlag(BuilderFlag::kINT8);
// layer->getOutput(0)->setType(nvinfer1::DataType::kINT8);
auto output = layer->getOutput(0);
network->markOutput(*output);

the output is: 
[5,61,43,100,1,62,-42,53,57,61,-12,-61,-80,62,-63,-90,19,62,56,-1,12,12,61,0,0,0,0,52,28,13,61,-16,46,110,62,0,0,0,0,-96,117,-19,60,114,25,-88,62,0,0,0,0,75,79,120,61,-35,-13,-53,61,-105,77,127,62,-105,5,-116,,60,0,0,0,0,7,-18,0,62,-34,48,62,60,-3,-17,-124,61,-121,5,21,59,119,90,-42,58,-26,50,58,60,0,0,0,0,23,125,-63,61,99,127,34,63,67,-45,33,59,-42,24,-54,61,0,0,0,0,0,0,0,0,-96,-70,-29,59,48,33,41,60,-5,48,35,61,-114,21,55,61,108,-44,48,58,-32,-88,-52,62,59,-36,-116,60,-40,-6,-56,62,88,9    2,-119,60,-3,87,-43,62,23,-86,-98,60,0,0,0,0,26,92,-110,59,75,-25,6,62,8,92,-66,60,0,0,0,0,96,-116,-10,61,22,-24,-73,62,63,25,42,61,-79,-61,14,62,-34,-86,-44,58,4,51,112,62,0,0,0,0,15,19,99,62,126,62,78,62,14,-41,42,61,108,110,-11,60,-23,-25,47,61,55,31,-9,60.....]
seems right?

**Comment by ttyio** [comment]
@Bonsen could you also add 
      config->setFlag(nvinfer1::BuilderFlag::kSTRICT_TYPES);
you can validate the result by multiple the scale back, and compare with the floating values.

**Comment by Bonsen** [comment]
@ttyio 
I set  config->setFlag(nvinfer1::BuilderFlag::kSTRICT_TYPES); but the results are same. 
It's strange the value is 0~128 when i set layer->getOutput(0)->setType(nvinfer1::DataType::kINT8);.

I set int8 with a calibration:
auto calibrator = new Int8CacheCalibrator(calibration_path);
config->setInt8Calibrator(calibrator);

The output layer use "StatefulPartitionedCall/inception_v3/global_average_pooling2d/Mean:0: 3c389938" in calibration file.
How can I get the scale to convert int8 to float for validation?
